### DISCUSSION TITLE
Running the same query twice results in four times the performance cost.

### DISCUSSION THREAD

**Discussion by thenbe** [prisma/prisma#11962] [Q&A] [answered]
This query takes **60ms** to execute:
```ts
prismaClient.$transaction([
    prismaClient.tenant.count(),
]);
```

This costs **250ms**!
```ts
prismaClient.$transaction([
    prismaClient.tenant.count(),
    prismaClient.tenant.count(),
]);
```

Obviously, I am not really counting twice, this is just to illustrate the issue in the simplest way. Is there a way to optimize this? Note that I do not really care about the `$transaction` guarantee that both queries either succeed or fail. The only reason I'm using it is to avoid make two round trips to the db. Is there an alternative I should use? If not, how can I optimize this? Even better, is there a way to run both queries asynchronously instead of sequentially? Any recommendations are welcome. Thank you.

Extra info: My actual use case is pagination. First query is `findMany`, second is `count`. This is in order to display:
_Showing 20 - 30 of total 140_ to my users.

Here is the debug output from both:

First Query (60ms):
```
{
  timestamp: 1645462618471,
  query: 'SELECT COUNT(*) FROM (SELECT `redb`.`Tenant`.`id` FROM `redb`.`Tenant` WHERE 1=1) AS `sub`',
  params: '[]',
  duration: '66',
  target: 'quaint::connector::metrics'
}
```

Second Query (250ms):
```
{
  timestamp: 1645462620703,
  query: 'BEGIN',
  params: '[]',
  duration: '59',
  target: 'quaint::connector::metrics'
}
{
  timestamp: 1645462620779,
  query: 'SELECT COUNT(*) FROM (SELECT `redb`.`Tenant`.`id` FROM `redb`.`Tenant` WHERE 1=1) AS `sub`',
  params: '[]',
  duration: '75',
  target: 'quaint::connector::metrics'
}
{
  timestamp: 1645462620849,
  query: 'SELECT COUNT(*) FROM (SELECT `redb`.`Tenant`.`id` FROM `redb`.`Tenant` WHERE 1=1) AS `sub`',
  params: '[]',
  duration: '69',
  target: 'quaint::connector::metrics'
}
{
  timestamp: 1645462620909,
  query: 'COMMIT',
  params: '[]',
  duration: '59',
  target: 'quaint::connector::metrics'
}
```

**Comment by millsp** [comment]
Hey @ambiguous48, were you using `interactiveTransactions`? If so, this should be fixed now in `3.10.0`. If not, then let's take a look at what could be wrong here. Thanks!

**Reply by thenbe** [reply]
@millsp Hey. Not using `interactiveTransactions`. Still getting the same issue on `3.10.0`.

**Reply by millsp** [reply]
Alright, then this is a question for the Engine team.

**Comment by millsp** [comment, marked as answer]
Sending a single query via `$transaction` does not open a transaction, but sending more than one adds the overhead of having to start one. That's the overhead caused by sending `BEGIN` and `COMMIT`, as well as the extra queries. Reducing the latency/distance between your app and your DB will help.

**Reply by thenbe** [reply]
@millsp This makes sense. Thank you!

**Reply by millsp** [reply]
You're welcome! Feel free to ping me whenever you need help. Cheers

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
